[doc] Add DIF Specification, Checklists, Progress

lenary opened the pull request:
This adds the agreed version of the "OpenTitan Device Interface
Functions" document. It is split between several different locations:

doc/rm/device_interface_functions.md - The main DIF document
doc/project/hw_stages.md - Added DIF stages and checklist reference.
doc/project/checklist.md - Added list of DIF checklist items

This change also includes a template for the DIF checklist:
doc/project/sw_checklist.md.tpl

--- 

There are also two other commits, one which adds the DIF status to the hardware
dashboard, and one which sets the three committed DIFs (uart, plic, gpio) to stage `S0`.
This allows the updated dashboard to be previewed, but not the checklist yet.

---

FAO @sjgitty and @gkelly so we can start discussions on how we want this to be structured, or whether more changes are required to integrate DIF stages into existing lifecycle documentation.

sjgitty commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,167 @@
+
+DIFs **must** have unit tests, which cover all their specification-required functionality.
+These unit tests use a mocking system which allows the test to control all hardware reads and writes, without requiring complex management of the hardware.
+These tests are written using googletest, and may be run on an x86 device.
```
Reference for `googletest`?

sjgitty commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,167 @@
+
+DIFs **should** also have other tests, including (but not limited to):
+
+*   Standalone sanity tests, which can quickly diagnose major issues between the DIF and the hardware, but are not required to test all functionality.
```
Since there is only one element here, does it make sense to make it a list?

lenary commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,167 @@
+The prose descriptions should include references to relevant DIF functions.
+
+Some of the programmers' guides also include standalone example code.
+This will be removed, as it is hard to keep up to date (we are no longer following the macro-based code style shown in existing examples) and introducing the `mmio_region_t` interface is not standalone.
```
Ok, I've rephrased this paragraph, with the emphasis on the API (doxygen) documentation, and suggesting (but not requiring) the avoidance of standalone examples. It's not exactly what you wrote, but I hope it gets the point across.

lenary commented on doc/project/hw_stages.md:
```diff
@@ -43,8 +43,10 @@ Feature requests towards a signed-off design requires review and approval by the
 Signed off fully-functioning (read: not buggy) designs stay in the Signoff stage as an available complete IP, with an associated revision ID.
-[Here](https://github.com/lowRISC/opentitan/blob/master/doc/project/ip_checklist.md.tpl) is a template that can be used as a checklist item.
 
+There exists a [template for IP checklists](https://github.com/lowRISC/opentitan/blob/master/doc/project/ip_checklist.md.tpl).
```
Will do, but probably in a separate commit, given that might have some fall-out in the rest of the documentation site.

lenary commented:
In recent updates:
- Addressed @sjgitty's feedback, including the file name change (which is in a separate commit)
- Requested reviews from the original list of suggested reviewers from the OT Standards discussion.

sjgitty commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,166 @@
+The prose descriptions should include references to relevant DIF functions.
+
+Programmers' guides **should** primarily refer to the [Software API Documentation]({{< relref "/sw#opentitan-software-api-documentation" >}}).
+Programmers' guides **should not** standalone examples that do not match how we implement DIFs and can become out of date.
```
"**should not**" `be` "standalone examples" perhaps?

lenary replied:
ah yes I have missed a word - i think 'contain' is the right word? This is the diff I propose.

```suggestion
Programmers' guides **should not** contain standalone examples that do not match how we implement DIFs and can become out of date.
```

sriyerg commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,166 @@
+
+DIFs **should** also have other tests, including standalone C sanity tests---which can quickly diagnose major issues between the DIF and the hardware but are not required to test all device functionality.
+
+DIFs are also being used by DV, which should help catch any issues with the DIFs not corresponding to the hardware implementation exactly.
```
These very likely won't live in `sw/device/tests/dif` folder as line 140 suggests.

lenary replied:
Ok, I have added a sentence that states this. 

At some point, it would be useful if you could choose a directory that we can put into this document, but I am happy for this to happen later.

lenary commented on doc/project/checklist.md:
```diff
@@ -395,3 +395,91 @@ Clean up all compile-time and run-time warnings thrown by the simulator.
+
+### DIF_TEST_UNIT_COMPLETE
+
+Unit test cover (at least):
```
Illegal Argument values are covered by "All DIF return codes". I have not yet had time to write up the guidance about DIF return codes because I'm trying to get this landed, but I have stated the guidance in a comment here: https://github.com/lowrisc/opentitan/issues/1911#issuecomment-613579290

lenary replied:
With regards to the phrase "Unit Test" – this is an accepted and specific software testing term.

lenary commented on doc/project/sw_checklist.md.tpl:
```diff
@@ -0,0 +1,74 @@
+from L0 (Specification) to L1 (Development) stage, and updated as needed.
+Once done, please remove this comment before checking it in.
+-->
+This checklist is for [Development Stage]({{< relref "/doc/project/development_stages.md" >}}) transitions for the [${name.upper()} DIF.]({{< relref "hw/ip/${name}/doc" >}})
```
Done.

lenary commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,166 @@
+### Coverage Requirements
+
+DIFs *must* actuate all of the specification-required functionality of the hardware that they are written for, and no more.
+A DIF cannot be declared complete until it provides an API for accessing all of the functionality that is expected of the hardware.
```
Correct.

lenary commented on util/dashboard/gen_dashboard_entry.py:
```diff
@@ -103,6 +111,34 @@ def get_linked_checklist(obj, rev, stage, is_latest_rev=True):
+    return "<a href=\"{}\">{}</a>".format(url, html.escape(rev[stage]))
+
 
 # Link development stages in "L# : D# : V#" format.
```
Done.

lenary commented on doc/project/checklist.md:
```diff
@@ -395,3 +395,91 @@ Clean up all compile-time and run-time warnings thrown by the simulator.
+
+### DIF_TEST_UNIT
+
+Unit tests exist for the DIF in `sw/device/tests/dif` named `dif_<ip>_unittest.cc`.
```
You are right that the current testing implementation is mock testing. However, @gkelly and I do not want to commit ourselves to mock testing forever, so we've phrased it as "unit tests", which is a software testing approach, rather than "mocked tests" which is an implementation technique.

lenary replied:
We don't have that documentation yet, but we can look to adding it in the future.

sriyerg commented on doc/project/checklist.md:
```diff
@@ -395,3 +395,91 @@ Clean up all compile-time and run-time warnings thrown by the simulator.
+
+### DIF_TEST_SANITY
+
+Sanity tests exist for the DIF in `sw/device/tests/dif` named `dif_<ip>_sanitytest.c`.
```
Is this a sanity for the DIF or for the hardware it targets? Does the DIF writer write this test, or is it a responsibility of the hardware / DV people? I'd imagine the sanity test would look different based on answers to these questions. We may want to ensure that we all are on the same page. The description IMO should clarify all this.

lenary replied:
I will add a clarifying sentence or two to point out:
- Hardware developers are already writing C tests for their main datapath, so they can debug them during development. There is no DIF for the module when these tests are written. For the moment, leaving these tests at this stage in `sw/device/tests/<ip>/<ip>_test.c` is fine.
- Later, when a DIF is written, these tests should be updated by the DIF author to use the DIF, and when they have been they will be put in `sw/device/tests/dif` named `dif_<ip>_sanitytest.c`.

sriyerg commented on doc/project/checklist.md:
```diff
@@ -395,3 +395,91 @@ Clean up all compile-time and run-time warnings thrown by the simulator.
+### DIF_TEST_SANITY
+
+Sanity tests exist for the DIF in `sw/device/tests/dif` named `dif_<ip>_sanitytest.c`.
+This should perform an end-to-end use of the hardware via the DIF, and should be able to be run on all OpenTitan platforms (including FPGA and simulation).
```
IMO an end-to-end test would also involve all other IPs the hardware interacts with (a system level scenario with multiple pieces of hardware that interact with each other and are working together towards a common goal). 

From your answer above, it is more like a "basic test run from the embedded core that exercises the main data path in the hardware for which the DIF is written".

lenary replied:
Point taken about end-to-end.

I'm happy to phrase this as "This should perform a basic test of the main datapath of the hardware module by the embedded core, via the DIF, and should be able to be run on all OpenTitan platforms (including FPGA, simulation and DV)."

I will also add "This test will be shared with DV".

sriyerg commented on doc/project/checklist.md:
```diff
@@ -395,3 +395,91 @@ Clean up all compile-time and run-time warnings thrown by the simulator.
+
+### DIF_HW_PARAMS
+
+DIF uses automatically generated HW parameters and register definitions.
```
I see, okay. Wouldn't the reggen / topgen stabilize at some point (probably very soon) once those extra definitions get added for auto-generation, so future DIFs wont have to hand-write anything at all? Just curious, fine with keeping this here in any case.

lenary replied:
Yes, we do envision reggen stabilising eventually.

I realise actually that topgen is less of an issue, as beyond parameterisations of peripherals (which are captured by reggen running on a topgen'd `.hjson` at the moment), because we're trying not to have any device-specific code in DIFs themselves. In effect, topgen is responsible for generating information that a user needs to use a DIF, but nothing internal to the DIF's implementation.

lenary commented on doc/project/checklist.md:
```diff
@@ -2,8 +2,8 @@
 
-This document explains the recommended checklist items to review when transitioning from one [Hardware Stage]({{<relref "/doc/project/hw_stages.md" >}}) to another, for both design and verification stages.
-It is expected that the items in each stage (D1, V1, etc) are completed.
+This document explains the recommended checklist items to review when transitioning from one [Development Stage]({{<relref "/doc/project/development_stages.md" >}}) to another, for design, verification, and [DIF]({{< relref "doc/rm/device_interface_functions.md" >}}) stages.
```
Oh this is done too.

lenary commented:
I think I have addressed all the points raised so far. I hope this is ready for sign-off and merge!

sjgitty commented:
Does this need further review, or is it ready to go?

asb commented on doc/project/checklist.md:
```diff
@@ -395,3 +395,102 @@ Clean up all compile-time and run-time warnings thrown by the simulator.
+
+### DIF_TEST_UNIT_COMPLETE
+
+Unit test cover (at least):
```
Nit: "Unit tests" (plural)

lenary replied:
Done.

asb commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,167 @@
+## Verification Stage Allowances
+
+DV, FPGA, and early silicon validation have unique requirements in their use of hardware.
+They might actuate hardware on vastly different timescales, use verification-specific registers, or otherwise manipulate aspects of the hardware that “production”-level code would not (or cannot) do.
```
Nit: fix non-ASCII quotes

lenary replied:
Done.

asb commented on doc/rm/device_interface_functions.md:
```diff
@@ -0,0 +1,167 @@
+
+In the hardware world, there are checklists for establishing that each stage is complete and the next stage can be moved to.
+
+In Software, we don’t have to follow the same stability guarantees, because software is much more modifiable.
```
Nit: "Software" seems randomly capitalised

lenary replied:
Done, and fixed the non-ascii apostrophe.

lenary commented:
In Today's Pushes:
- Rebased onto master
- Addressed @asb's feedback: removed non-ascii quotes and apostrophes, and typos.
- Added an alias from hw_stages to development_stages.

asb commented:
Thanks Sam - this has been reviewed by many people now, and I understand that Garret has indicated his approval off-line, so I went ahead and landed this.
